### PR TITLE
Dev

### PR DESCRIPTION
**Pull request by Gelelus**

### CODE REVIEW

File: src/app/profile/profile-routing.module.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+
+const routes: Routes = [
+    {
+      path: '',
+      component: ProfileComponent,
+      canActivate: [AuthGuard],
+      children: [
+        { path: '', component: SettingComponent },
+        { path: 'setting', component: SettingComponent },
```
Comment (ZubelRuslan) [inline src/app/profile/profile-routing.module.ts:17]: Is it temporarily implementation for testing

File: src/app/profile/profile.component.ts [hunk 2]
```diff
@@ -1,19 +1,42 @@
...
+
+  constructor(private store: Store<fromApp.AppState>) {}
 
-  ngOnInit() {}
+  ngOnInit() {
+    this.userSub = this.store
+      .pipe(
+        select('auth'),
+        map((authState) => authState.user)
+      )
```
Comment (ZubelRuslan) [inline src/app/profile/profile.component.ts:29]: Create specific selector

File: src/app/profile/profile.component.ts [hunk 3]
```diff
@@ -1,19 +1,42 @@
...
+      .subscribe((user) => {
+        if (user) {
+          
+          this.imgAvatarUrl = environment.DataBaseUrl +  user.avatarImgUrl;
+          this.name = user.name
+          this.registratedDate = new Date(user.date).toISOString().slice(0, 16).replace('T',' ')
```
Comment (ZubelRuslan) [inline src/app/profile/profile.component.ts:35]: You can remove `subscribe` and use pipe `async` in template

File: src/app/profile/setting/setting.component.ts [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+  private initForm() {
+    let firstName = '';
+    let secondName = '';
+    let email = '';
+    let phoneNumber = '';
+
+    this.userSub = this.store
+      .pipe(
+        select('auth'),
+        map((authState) => authState.user)
```
Comment (ZubelRuslan) [inline src/app/profile/setting/setting.component.ts:34]: create specific selector

File: src/app/profile/setting/setting.component.ts [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+          this.profileForm = new FormGroup({
+            firstName: new FormControl(firstName),
+            secondName: new FormControl(secondName),
+            email: new FormControl({value: email, disabled: true}),
+            phoneNumber: new FormControl(phoneNumber),
+            password: new FormControl(null),
+            secondPassword: new FormControl(null),
+          });
+        }
```
Comment (ZubelRuslan) [inline src/app/profile/setting/setting.component.ts:52]: You can do this in `tap` operator in `pipe` section and you should use `setValue` method for existing controls

File: src/app/recipes/recipe-edit/recipe-edit.component.html [hunk 6]
```diff
@@ -5,12 +5,13 @@
       <div class="row">
         <div class="col-xs-12">
           <button
-            [disabled]="!recipeForm.valid"
+            [disabled]="!(recipeForm.valid || EditImage)"
```
Comment (ZubelRuslan) [inline src/app/recipes/recipe-edit/recipe-edit.component.html:8]: please starts your variables names from lower case

File: src/app/recipes/recipe-edit/recipe-edit.component.ts [hunk 7]
```diff
@@ -72,21 +88,17 @@ export class RecipeEditComponent implements OnInit, OnDestroy {
...
-        .pipe(
-          select('recipes'),
-          map((recipeState) => {
-            return recipeState.recipes.find((recipe) => recipe._id === this.id);
-          })
-        )
+        .pipe(select(RecipesSelectors.findRecipeById, { id: this.id }))
         .subscribe((recipe) => {
+          this.imagePreview = environment.DataBaseUrl + recipe.imagePath;
+          this.EditImage = true;
```
Comment (ZubelRuslan) [inline src/app/recipes/recipe-edit/recipe-edit.component.ts:100]: Please combine rxjs operators instead `subscribe`. Here you can use `map` operator after `select` in `pipe`. In map you will toggle `EditImage` flag and return receipt, that you will store in class variable. In template you can use pipe `async` to show data

File: src/app/recipes/recipe-list/recipe-item/recipe-item.component.ts [hunk 8]
```diff
@@ -8,7 +10,9 @@ import { Recipe } from '../../recipe.model';
 })
 export class RecipeItemComponent implements OnInit {
   @Input() recipe: Recipe;
+  ServerUrl = environment.DataBaseUrl;
```
Comment (ZubelRuslan) [inline src/app/recipes/recipe-list/recipe-item/recipe-item.component.ts:13]: `serverUrl`

File: src/app/recipes/recipe-list/recipe-list.component.ts [hunk 9]
```diff
@@ -27,14 +28,16 @@ export class RecipeListComponent implements OnInit, OnDestroy {
...
-          return recipesState.recipes;
-        })
-      )
-      .subscribe((recipes: Recipe[]) => {
-        this.recipes = recipes;
+      .pipe(select('recipes'))
+      .subscribe((recipesState) => {
+        this.recipes = recipesState.recipes;
+        this.recipesNumber = recipesState.maxRecipes;
+        this.recipesOnPage = recipesState.recipesOnPage;
```
Comment (ZubelRuslan) [inline src/app/recipes/recipe-list/recipe-list.component.ts:35]: Please use specific selectors

File: src/app/recipes/recipe-list/recipe-list.component.ts [hunk 10]
```diff
@@ -53,7 +55,9 @@ export class RecipeListComponent implements OnInit, OnDestroy {
           },
         })
         .then(() => {
-          this.store.dispatch(new RecipeActions.FetchRecipes());
+          this.store.dispatch(
+            new RecipeActions.FetchRecipes({ startItem: 0, limit: 5 })
```
Comment (ZubelRuslan) [inline src/app/recipes/recipe-list/recipe-list.component.ts:59]: This independent from current page url?

File: src/app/recipes/recipes.component.ts [hunk 11]
```diff
@@ -1,12 +1,21 @@
...
   selector: 'app-recipes',
   templateUrl: './recipes.component.html',
   styleUrls: ['./recipes.component.css'],
 })
-export class RecipesComponent implements OnInit {
-  constructor() {}
-
-  ngOnInit(): void {}
+export class RecipesComponent implements OnInit{
+  error: Observable<string> = null;
```
Comment (ZubelRuslan) [inline src/app/recipes/recipes.component.ts:14]: you can immediately initialize this variable `error: Observable<string> = this.store.pipe(select(RecipesSelectors.error))`

File: src/app/recipes/store/recipe.actions.ts [hunk 12]
```diff
@@ -29,17 +32,32 @@ export class AddRecipe implements Action {
...
   readonly type = UPDATE_RECIPE_ON_DB;
-  constructor(public payload: Recipe ) {}
+  constructor(
+    public payload: {
+      description: string;
+      image: File;
+      ingredients: Ingredient[];
+      name: string;
+      _id: string;
+    }
```
Comment (ZubelRuslan) [inline src/app/recipes/store/recipe.actions.ts:59]: You don't use this type anywhere else?

File: src/app/recipes/store/recipe.effect.ts [hunk 13]
```diff
@@ -1,41 +1,55 @@
...
       if (url.includes('?')) {
         httpParams = new HttpParams({ fromString: url.split('?')[1] });
       }
-      return this.http.get<Recipe[]>(environment.DataBaseUrl + 'recipes', {
-        params: httpParams,
-      });
+      return this.http.get<{ recipes: Recipe[]; maxRecipes: number }>(
+        environment.DataBaseUrl + 'recipes',
+        {
+          params: httpParams,
```
Comment (ZubelRuslan) [inline src/app/recipes/store/recipe.effect.ts:36]: Why you didn't use simple params config object? `{ params: {startItem: actionData.payload.startItem, limit: actionData.payload.startItem}}`

File: src/app/recipes/store/recipe.effect.ts [hunk 14]
```diff
@@ -1,41 +1,55 @@
...
+    map((postData) => {
+      return {
+        recipes: postData.recipes.map((recipe) => {
+          return {
+            ...recipe,
+            ingredients: recipe.ingredients ? recipe.ingredients : [],
+          };
+        }),
+        maxRecipes: postData.maxRecipes,
+      };
```
Comment (ZubelRuslan) [inline src/app/recipes/store/recipe.effect.ts:49]: As I see there no transformations. For what you implement this `map`?

**Review by ZubelRuslan — CHANGES_REQUESTED** [review]

File: src/app/profile/profile-routing.module.ts (hunk 1 above)
Comment (Gelelus) [reply src/app/profile/profile-routing.module.ts:17]: yes

File: src/app/profile/profile.component.ts (hunk 2 above)
Comment (Gelelus) [reply src/app/profile/profile.component.ts:29]: created already

File: src/app/profile/setting/setting.component.ts (hunk 4 above)
Comment (Gelelus) [reply src/app/profile/setting/setting.component.ts:34]: created already

File: src/app/recipes/recipe-edit/recipe-edit.component.html (hunk 6 above)
Comment (Gelelus) [reply src/app/recipes/recipe-edit/recipe-edit.component.html:8]: ok

File: src/app/recipes/recipe-list/recipe-item/recipe-item.component.ts (hunk 8 above)
Comment (Gelelus) [reply src/app/recipes/recipe-list/recipe-item/recipe-item.component.ts:13]: ups

File: src/app/profile/profile.component.ts (hunk 3 above)
Comment (Gelelus) [reply src/app/profile/profile.component.ts:35]: subscribe removed, used pipe async

File: src/app/profile/setting/setting.component.ts (hunk 5 above)
Comment (Gelelus) [reply src/app/profile/setting/setting.component.ts:52]: fixed, but i use patchValue method.It's ok?

File: src/app/recipes/recipe-edit/recipe-edit.component.ts (hunk 7 above)
Comment (Gelelus) [reply src/app/recipes/recipe-edit/recipe-edit.component.ts:100]: fixed

File: src/app/recipes/recipe-list/recipe-list.component.ts (hunk 9 above)
Comment (Gelelus) [reply src/app/recipes/recipe-list/recipe-list.component.ts:35]: added specific selectors and pipe `async`

File: src/app/recipes/recipe-list/recipe-list.component.ts (hunk 10 above)
Comment (Gelelus) [reply src/app/recipes/recipe-list/recipe-list.component.ts:59]: fixed

File: src/app/recipes/recipes.component.ts (hunk 11 above)
Comment (Gelelus) [reply src/app/recipes/recipes.component.ts:14]: fixed

File: src/app/recipes/store/recipe.actions.ts (hunk 12 above)
Comment (Gelelus) [reply src/app/recipes/store/recipe.actions.ts:59]: replaced

File: src/app/recipes/store/recipe.effect.ts (hunk 13 above)
Comment (Gelelus) [reply src/app/recipes/store/recipe.effect.ts:36]: fixed

File: src/app/recipes/store/recipe.effect.ts (hunk 14 above)
Comment (Gelelus) [reply src/app/recipes/store/recipe.effect.ts:49]: Fixed. Yes, now it is not needed. I just used firebase like server at the beginning, there were problems with this ingredients.

**Pull request merged**